src/components/core/base/BaseVisualization.tsx

```diff
@@ -0,0 +1,159 @@
+import * as React from 'react';
+import noop = require('lodash/noop');
+import isEqual = require('lodash/isEqual');
+import {
+    DataSource,
+    DataSourceUtils,
+    createSubject
+} from '@gooddata/data-layer';
+import { AFM, Execution } from '@gooddata/typings';
+
+import { IEvents, ILoadingState } from '../../../interfaces/Events';
+import { IDrillableItem } from '../../../interfaces/DrillEvents';
+
+import { ErrorStates } from '../../../constants/errorStates';
+import { getVisualizationOptions } from '../../../helpers/options';
+import { convertErrors, checkEmptyResult } from '../../../helpers/errorHandlers';
+import { ISubject } from '../../../helpers/async';
+
+export type IHeadlineDataPromise = Promise<Execution.IExecutionResponses>;
+
+export interface IBaseVisualizationProps extends IEvents {
+    dataSource: DataSource.IDataSource<Execution.IExecutionResponses>;
+    resultSpec?: AFM.IResultSpec;
+    locale?: string;
+    drillableItems?: IDrillableItem[];
+    afterRender?: Function;
+    pushData?: Function;
+    ErrorComponent?: React.ComponentClass<any>;
+    LoadingComponent?: React.ComponentClass<any>;
+}
+
+export interface IBaseVisualizationState {
+    error: string;
+    result: Execution.IExecutionResponses;
+    isLoading: boolean;
+}
+
+export abstract class BaseVisualization<P extends IBaseVisualizationProps, S extends IBaseVisualizationState>
+    extends React.Component<P, S> {
+
+    protected subject: ISubject<IHeadlineDataPromise>;
+
+    constructor(props: P) {
+        super(props);
+
+        this.onLoadingChanged = this.onLoadingChanged.bind(this);
+        this.onDataTooLarge = this.onDataTooLarge.bind(this);
+        this.onError = this.onError.bind(this);
+
+        this.initSubject();
+    }
+
+    public componentWillMount() {
+        this.setState({
+            error: ErrorStates.OK,
+            result: null,
+            isLoading: false
+        });
+    }
+
+    public componentDidMount() {
+        const { dataSource, resultSpec } = this.props;
+        this.initDataLoading(dataSource, resultSpec);
+    }
+
+    public render() {
+        const { result, isLoading, error } = this.state;
+        const { ErrorComponent, LoadingComponent } = this.props;
+
+        if (error !== ErrorStates.OK) {
+            return ErrorComponent ? <ErrorComponent error={{ status: error }} props={this.props} /> : null;
```

Comment:
Is it OK that nothing will be rendered? Shouldn't we use some default error component with a generic message? Maybe for loading below as well?